### PR TITLE
[MS-230] Add parser modes feature

### PR DESCRIPTION
**Pull request by varsill**
This PR introduces three modes the parser can work in:
* `:bytestream`
* `:nalu_aligned`
* `:au_aligned`
The names of the modes are connected with the structure of input buffers.

### CODE REVIEW

File: lib/membrane_h264_plugin/parser.ex [hunk 1]
```diff
@@ -54,24 +58,45 @@ defmodule Membrane.H264.Parser do
...
+
+    state = %{state | mode: mode}
     {:ok, state}
   end
 
   @impl true
   def handle_process(:input, %Membrane.Buffer{} = buffer, _ctx, state) do
-    payload = state.params <> buffer.payload
-    {nalus_payloads_list, nalu_splitter} = NALuSplitter.split(payload, state.nalu_splitter)
+    {nalus_payloads_list, nalu_splitter} = NALuSplitter.split(buffer.payload, state.nalu_splitter)
```
Comment (mickel8) [inline lib/membrane_h264_plugin/parser.ex:86]: We might consider not calling this in nalu_aligned mode but I don't insist

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: lib/membrane_h264_plugin/parser/au_splitter.ex [hunk 2]
```diff
@@ -134,7 +134,8 @@ defmodule Membrane.H264.Parser.AUSplitter do
   end
 
   def split_nalus([], state) do
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_h264_plugin/parser/au_splitter.ex:136]: done

File: test/integration/modes_test.exs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+  @h264_input_file "test/fixtures/input-10-720p.h264"
+  defp prepare_buffers(binary, mode) when mode == :bytestream do
+    buffers =
+      :binary.bin_to_list(binary) |> Enum.chunk_every(400) |> Enum.map(&:binary.list_to_bin(&1))
+
+    Enum.map(buffers, &%Membrane.Buffer{payload: &1})
+  end
+
+  defp prepare_buffers(binary, mode) when mode == :nalu_aligned do
```
[the remark this answers is not in the document]
Comment (varsill) [reply test/integration/modes_test.exs:20]: Well, I am keen on that syntax as it now the name of the variable (`:mode`) - but I also know that it is probably no adivsed, so I have changed it ;)
done

File: test/integration/modes_test.exs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    Enum.map_reduce(aus, 0, fn au, ts ->
+      {for(nalu <- au, do: %Membrane.Buffer{payload: nalu.payload, pts: ts, dts: ts}), ts + 1}
+    end)
+    |> elem(0)
+    |> Enum.flat_map(& &1)
+  end
+
+  defp prepare_buffers(binary, mode) when mode == :au_aligned do
+    {nalus_payloads, nalu_splitter} = binary |> NALuSplitter.split(NALuSplitter.new())
```
[the remark this answers is not in the document]
Comment (varsill) [reply test/integration/modes_test.exs:40]: done

File: lib/membrane_h264_plugin/parser.ex (hunk 1 above)
Comment (varsill) [reply lib/membrane_h264_plugin/parser.ex:86]: I will leave it that way and we will change it with the introduction of input and output behaviours, as we discussed yesterday

File: lib/membrane_h264_plugin/parser/au_splitter.ex [hunk 5]
```diff
@@ -52,7 +52,7 @@ defmodule Membrane.H264.Parser.AUSplitter do
 
   @type access_unit_t() :: list(NALu.t())
 
-  # split_nalus/5 defines a finite state machine with two states: :first and :second.
+  # split/5 defines a finite state machine with two states: :first and :second.
   # The state :first describes the state before reaching the primary coded picture NALu of a given access unit.
   # The state :second describes the state after processing the primary coded picture NALu of a given access unit.
```
Comment (DominikWolek) [inline lib/membrane_h264_plugin/parser/au_splitter.ex:57]: I know this is not the main topic of the PR, but these line are worth including into the documentation - perhaps moving description of `:first` and `:second` to the typedoc of `t()`

**Review by DominikWolek — APPROVED** [review]

File: lib/membrane_h264_plugin/parser/au_splitter.ex (hunk 5 above)
Comment (varsill) [reply lib/membrane_h264_plugin/parser/au_splitter.ex:57]: I am not 100% convinced - `AUSplitter.t()` is opaque and I believe that the knowledge of what does particular state represent is worth only for those who are reading the implementation of the `split/2` function - therefore, as far as I am concerned, I wouldn't include that information in the public documentation.
WDYT @mickel8 ?

**Pull request merged**